File: tests/current/test_domain_golden.py
"""Execute the shared domain vectors against the real Python runtime helpers."""
from __future__ import annotations

import json
import sys
from datetime import date
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))

from backend import content  # noqa: E402


FIXTURE = json.loads((ROOT / "contracts" / "domain-golden-v1.json").read_text(encoding="utf-8"))
DIFFICULTIES = ("easy", "medium", "hard", "hardcore", "mozkomor")
POINTS = {"daily": 100, "easy": 15, "medium": 25, "hard": 50, "hardcore": 100, "mozkomor": 150}


def test_xp_vectors() -> None:
    for vector in FIXTURE["xp"]:
        assert content.xp_for(
            vector["mode"], vector.get("difficulty"), POINTS,
            reward_xp=vector.get("rewardXp"),
        ) == vector["expected"]


def test_challenge_key_vectors() -> None:
    for vector in FIXTURE["challengeKeys"]:
        assert content.challenge_key(
            vector["mode"], vector["puzzleId"], vector.get("dailyDate"),
        ) == vector["expected"]


def test_streak_vectors() -> None:
    for vector in FIXTURE["streak"]:
        current, longest = content.streaks(vector["dates"], date.fromisoformat(vector["today"]))
        assert (current, longest) == (vector["current"], vector["longest"])


def test_unlock_vectors() -> None:
    for vector in FIXTURE["unlock"]:
        slots = {"baseCurrent": {"hardcore": vector["baseCurrentHardcore"]}}
        assert content.mozkomor_unlocked_from_rows(vector["rows"], slots) is vector["expected"]


def test_rank_vectors() -> None:
    rows = [vector["serverRow"] for vector in FIXTURE["rank"]]
    assert [list(content.run_rank_tuple(row)) for row in rows] == [vector["expected"] for vector in FIXTURE["rank"]]
    assert content.competition_ranks(rows) == [1, 1, 3, 4]


def test_daily_vectors() -> None:
    for vector in FIXTURE["daily"]:
        assert content.expected_daily_puzzle_id(vector["data"], vector["date"]) == vector["expected"]


def test_free_selection_vectors() -> None:
    selection = FIXTURE["freeSelection"]
    for vector in selection["cases"]:
        resolved = content.free_puzzle_info(
            selection["data"], selection["rolling"], vector["id"], DIFFICULTIES,
        )
        expected = vector["expected"]
        if expected is None:
            assert resolved is None
        else:
            assert {key: resolved.get(key) for key in expected} == expected


def test_release_vectors() -> None:
    for vector in FIXTURE["release"]:
        puzzle = {"meta": {"availableFrom": vector["availableFrom"]}}
        assert content.is_puzzle_released(puzzle, date.fromisoformat(vector["asOf"])) is vector["expected"]


def test_intentional_content_source_delta() -> None:
    server = json.loads((ROOT / "data" / "puzzles.json").read_text(encoding="utf-8"))
    public = json.loads((ROOT / "public" / "puzzles.json").read_text(encoding="utf-8"))
    assert len(server["daily"]) == FIXTURE["contentSources"]["serverDailyCount"]
    assert len(public["daily"]) == FIXTURE["contentSources"]["publicDailyCount"]


if __name__ == "__main__":
    tests = [value for name, value in sorted(globals().items()) if name.startswith("test_") and callable(value)]
    for test in tests:
        test()
    print(f"domain golden vectors (Python): {len(tests)} PASS")

File: tests/current/test_domain_golden.js
/* Execute the shared vectors against function bodies extracted from public/app.js. */
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

const root = path.join(__dirname, '../..');
const fixture = JSON.parse(fs.readFileSync(path.join(root, 'contracts/domain-golden-v1.json'), 'utf8'));
const appSource = fs.readFileSync(path.join(root, 'public/app.js'), 'utf8');

function between(name, nextName) {
  const start = appSource.indexOf(`function ${name}(`);
  const end = appSource.indexOf(`function ${nextName}(`, start + 1);
  assert.notEqual(start, -1, `public/app.js function missing: ${name}`);
  assert.notEqual(end, -1, `public/app.js boundary missing after: ${name}`);
  return appSource.slice(start, end);
}

const storage = new Map();
const context = {
  console,
  Date,
  Set,
  Math,
  Number,
  Object,
  JSON,
  DIFF: {
    easy: {xp: 15}, medium: {xp: 25}, hard: {xp: 50},
    hardcore: {xp: 100}, mozkomor: {xp: 150},
  },
  TAJENKA_REWARD_XP: 200,
  MOZKOMOR_UNLOCK_BASE: 200,
  MOZKOMOR_QA_PREVIEW: false,
  MOZKOMOR_UNLOCK_KEY: 'golden-unlock',
  testToday: '2026-08-30',
  testTajenkaCompleted: false,
  testSlots: {actual: new Set()},
  testRemote: {},
  puzzleDB: null,
  pragueDateISO() { return context.testToday; },
  tajenkaCompletion() { return context.testTajenkaCompleted; },
  localFreeSlotState() { return context.testSlots; },
  getProfile() { return {stats: context.testRemote}; },
  scopedStorageKey(key) { return key; },
  localStorage: {
    getItem(key) { return storage.has(key) ? storage.get(key) : null; },
    setItem(key, value) { storage.set(key, String(value)); },
  },
};
vm.createContext(context);
const runtimeFunctions = [
  between('resultRankTuple', 'betterResult'),
  between('freePuzzleSlot', 'localFreeSlotState'),
  between('dayOffsetISO', 'dailyBankFor'),
  between('challengeKey', 'pointsFor'),
  between('pointsFor', 'savedProgressFor'),
  between('isoShift', 'streakEndingOn'),
  between('streakEndingOn', 'localRescueStatus'),
  between('calcStreak', 'calcLongest'),
  between('calcLongest', 'levelFor'),
  between('localMozkomorBaseDone', 'mozkomorUnlockState'),
  between('mozkomorUnlockState', 'renderQuickPlay'),
].join('\n');
vm.runInContext(`${runtimeFunctions}\nthis.runtime = {resultRankTuple, freePuzzleSlot, challengeKey, pointsFor, streakEndingOn, calcStreak, calcLongest, mozkomorUnlockState};`, context);
const runtime = context.runtime;
const progression=require(path.join(root,'public/app/content/progression.js')).create({
  getPuzzleDB:()=>context.puzzleDB,
  getState:()=>({completed:{}}),
  dayOffsetISO:(...args)=>context.dayOffsetISO(...args),
});
runtime.dailyPuzzleFor=iso=>progression.dailyPuzzleFor(iso);

context.puzzleDB = {
  freeGeneration: 4,
  free: {
    easy: [{id: 'easy-1', meta: {level: 1, contentGeneration: 4}}],
    hardcore: [{id: 'hardcore-1', meta: {level: 1, contentGeneration: 4}}],
    mozkomor: [{id: 'mozkomor-1', meta: {level: 1, contentGeneration: 4}}],
  },
};
for (const vector of fixture.xp) {
  context.testTajenkaCompleted = false;
  context.testSlots = {actual: new Set()};
  const puzzle = {id: vector.puzzleId || 'daily', meta: {rewardXp: vector.rewardXp}};
  assert.equal(runtime.pointsFor(vector.mode, vector.difficulty, puzzle), vector.expected);
}

for (const vector of fixture.challengeKeys) {
  assert.equal(runtime.challengeKey(vector.mode, {id: vector.puzzleId}, vector.dailyDate), vector.expected);
}

for (const vector of fixture.streak) {
  context.testToday = vector.today;
  assert.equal(runtime.calcStreak(vector.dates), vector.current);
  assert.equal(runtime.calcLongest(vector.dates), vector.longest);
}

for (const vector of fixture.unlock) {
  storage.clear();
  context.testRemote = {
    freeBasePlayedCurrent: {hardcore: vector.baseCurrentHardcore},
    mozkomorUnlocked: vector.rows.some(row => row.mode === 'free' && row.difficulty === 'mozkomor'),
  };
  context.testSlots = {actual: new Set()};
  assert.equal(runtime.mozkomorUnlockState().unlocked, vector.expected);
}

assert.deepEqual(
  fixture.rank.map(vector => Array.from(runtime.resultRankTuple(vector.clientRow))),
  fixture.rank.map(vector => vector.expected),
);

for (const vector of fixture.daily) {
  context.puzzleDB = vector.data;
  assert.equal(runtime.dailyPuzzleFor(vector.date).id, vector.expected);
}

context.puzzleDB = fixture.freeSelection.data;
for (const vector of fixture.freeSelection.cases) {
  const resolved = runtime.freePuzzleSlot(vector.id);
  if (vector.expected === null) {
    assert.equal(resolved, null);
  } else {
    const projected = {};
    for (const key of Object.keys(vector.expected)) projected[key] = resolved[key];
    assert.deepEqual(projected, vector.expected);
  }
}

const serverContent = JSON.parse(fs.readFileSync(path.join(root, 'data/puzzles.json'), 'utf8'));
const publicContent = JSON.parse(fs.readFileSync(path.join(root, 'public/puzzles.json'), 'utf8'));
assert.equal(serverContent.daily.length, fixture.contentSources.serverDailyCount);
assert.equal(publicContent.daily.length, fixture.contentSources.publicDailyCount);

console.log('domain golden vectors (actual runtime owners): PASS');
